### PR TITLE
feat(MessageBar): Initial implementation

### PR DESCRIPTION
**Pull request by ling1726**
Adds an initial implementation of MessageBar that includes design for different intents and multiline handling.

image

image

Addresses: #22579

### CODE REVIEW

**Comment by fabricteam** [comment]
## 📊 Bundle size report

<sub>🤖 This report was generated against <a href='https://github.com/microsoft/fluentui/commit/bb6155adb0828619519ec1a871d9628acf3b9a38'>bb6155adb0828619519ec1a871d9628acf3b9a38</a></sub>
<div id="bundleSizeComment9423"/>

**Review by bsunderhus — APPROVED** [review]

**Comment by fabricteam** [comment]
## [Perf Analysis (`@fluentui/react-components`)](https://github.com/microsoft/fluentui/wiki/Perf-Testing)
<p>No significant results to display.</p><details><summary>All results</summary><p>
  <table>
  <tr>
    <th>Scenario</th>
    <th>Render type</th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">Master Ticks</a>
    </th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">PR Ticks</a>
    </th>
    <th>Iterations</th>
    <th>Status</th>
  </tr><tr>
            <td>Avatar</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29313/perf-test-react-components/Avatar-mount_base.html">634</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29313/perf-test-react-components/Avatar-mount.html">616</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>Button</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29313/perf-test-react-components/Button-mount_base.html">309</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29313/perf-test-react-components/Button-mount.html">315</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>Field</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29313/perf-test-react-components/Field-mount_base.html">1111</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29313/perf-test-react-components/Field-mount.html">1111</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProvider</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29313/perf-test-react-components/FluentProvider-mount_base.html">680</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29313/perf-test-react-components/FluentProvider-mount.html">691</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29313/perf-test-react-components/FluentProviderWithTheme-mount_base.html">78</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29313/perf-test-react-components/FluentProviderWithTheme-mount.html">85</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>virtual-rerender</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29313/perf-test-react-components/FluentProviderWithTheme-virtual-rerender_base.html">65</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29313/perf-test-react-components/FluentProviderWithTheme-virtual-rerender.html">64</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>virtual-rerender-with-unmount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29313/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount_base.html">74</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29313/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount.html">71</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>InfoButton</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29313/perf-test-react-components/InfoButton-mount_base.html">10</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29313/perf-test-react-components/InfoButton-mount.html">10</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>MakeStyles</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29313/perf-test-react-components/MakeStyles-mount_base.html">856</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29313/perf-test-react-components/MakeStyles-mount.html">858</a></td>
            <td>50000</td>
            <td></td>
           </tr>
<tr>
            <td>Persona</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29313/perf-test-react-components/Persona-mount_base.html">1775</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29313/perf-test-react-components/Persona-mount.html">1677</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>SpinButton</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29313/perf-test-react-components/SpinButton-mount_base.html">1365</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/29313/perf-test-react-components/SpinButton-mount.html">1370</a></td>
            <td>5000</td>
            <td></td>
           </tr></table></p></details>

<div id="perfComment9425"/>

**Comment by fabricteam** [comment]
## 🕵 fluentuiv9 No visual regressions between this PR and main
<div id="vrtCommentfluent-ui_VRT_Pipeline_v9"/>

File: packages/react-components/react-message-bar-preview/src/components/MessageBar/MessageBar.types.ts [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+export type MessageBarContextValues = {
+  messageBar: MessageBarContextValue;
+};
+
+/**
+ * MessageBar Props
+ */
+export type MessageBarProps = ComponentProps<MessageBarSlots> &
+  Pick<MessageBarContextValue, 'layout'> & {
+    multiline?: boolean;
```
Comment (layershifter) [inline packages/react-components/react-message-bar-preview/src/components/MessageBar/MessageBar.types.ts:18]: Cruft?

File: packages/react-components/react-message-bar-preview/src/components/MessageBarActions/MessageBarActions.test.tsx [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+import { isConformant } from '../../testing/isConformant';
+import { MessageBarActions } from './MessageBarActions';
+import { MessageBarActionsProps } from './MessageBarActions.types';
+
+describe('MessageBarActions', () => {
+  isConformant<MessageBarActionsProps>({
+    Component: MessageBarActions,
+    displayName: 'MessageBarActions',
+    disabledTests: [
+      // TODO: having problems due to the fact root of DialogTitle is Fragment
```
Comment (layershifter) [inline packages/react-components/react-message-bar-preview/src/components/MessageBarActions/MessageBarActions.test.tsx:12]: What?

File: packages/react-components/react-message-bar-preview/src/components/MessageBarActions/MessageBarActions.tsx [hunk 3]
```diff
@@ -0,0 +1,18 @@
+import * as React from 'react';
+import type { ForwardRefComponent } from '@fluentui/react-utilities';
+import { useMessageBarActions_unstable } from './useMessageBarActions';
+import { renderMessageBarActions_unstable } from './renderMessageBarActions';
+import { useMessageBarActionsStyles_unstable } from './useMessageBarActionsStyles.styles';
+import type { MessageBarActionsProps } from './MessageBarActions.types';
+
+/**
+ * MessageBarActions component - TODO: add more docs
```
Comment (layershifter) [inline packages/react-components/react-message-bar-preview/src/components/MessageBarActions/MessageBarActions.tsx:9]: Let's solve todo

File: packages/react-components/react-message-bar-preview/src/components/MessageBar/MessageBar.types.ts (hunk 1 above)
Comment (ling1726) [reply packages/react-components/react-message-bar-preview/src/components/MessageBar/MessageBar.types.ts:18]: yeah removed in #29328

File: packages/react-components/react-message-bar-preview/src/components/MessageBarActions/MessageBarActions.types.ts [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+
+export type MessageBarActionsSlots = {
+  root: Slot<'div'>;
+  containerAction?: Slot<'div'>;
+};
+
+/**
+ * MessageBarActions Props
+ */
+export type MessageBarActionsProps = ComponentProps<MessageBarActionsSlots> & {};
```
Comment (layershifter) [inline packages/react-components/react-message-bar-preview/src/components/MessageBarActions/MessageBarActions.types.ts:12]: Let's remove `{}` as it's `any`

File: packages/react-components/react-message-bar-preview/src/components/MessageBarActions/MessageBarActions.test.tsx (hunk 2 above)
Comment (ling1726) [reply packages/react-components/react-message-bar-preview/src/components/MessageBarActions/MessageBarActions.test.tsx:12]: I copied that config from DialogTitle, will remove in another PR - the test itself should still be disabled tho

File: packages/react-components/react-message-bar-preview/src/components/MessageBarActions/renderMessageBarActions.tsx [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+import { ButtonContextProvider } from '@fluentui/react-button';
+
+/**
+ * Render the final JSX of MessageBarActions
+ */
+export const renderMessageBarActions_unstable = (state: MessageBarActionsState) => {
+  assertSlots<MessageBarActionsSlots>(state);
+  if (state.layout === 'multiline') {
+    return (
+      <ButtonContextProvider value={{ size: 'small' }}>
```
Comment (layershifter) [inline packages/react-components/react-message-bar-preview/src/components/MessageBarActions/renderMessageBarActions.tsx:15]: - This should be a memoized value as `ButtonContextProvider` is not using context selectors
- The value should be defined in `useMessageBarContextValues`

File: packages/react-components/react-message-bar-preview/src/components/MessageBarBody/MessageBarBody.types.ts [hunk 6]
```diff
@@ -0,0 +1,15 @@
+import type { ComponentProps, ComponentState, Slot } from '@fluentui/react-utilities';
+
+export type MessageBarBodySlots = {
+  root: Slot<'div'>;
+};
+
+/**
+ * MessageBarBody Props
+ */
+export type MessageBarBodyProps = ComponentProps<MessageBarBodySlots> & {};
```
Comment (layershifter) [inline packages/react-components/react-message-bar-preview/src/components/MessageBarBody/MessageBarBody.types.ts:10]: Let's remove `{}`

File: packages/react-components/react-message-bar-preview/src/components/MessageBarTitle/MessageBarTitle.types.ts [hunk 7]
```diff
@@ -0,0 +1,15 @@
+import type { ComponentProps, ComponentState, Slot } from '@fluentui/react-utilities';
+
+export type MessageBarTitleSlots = {
+  root: Slot<'span'>;
+};
+
+/**
+ * MessageBarTitle Props
+ */
+export type MessageBarTitleProps = ComponentProps<MessageBarTitleSlots> & {};
```
Comment (layershifter) [inline packages/react-components/react-message-bar-preview/src/components/MessageBarTitle/MessageBarTitle.types.ts:10]: Let's remove `{}`

File: packages/react-components/react-message-bar-preview/src/components/MessageBarTitle/renderMessageBarTitle.tsx [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+import { assertSlots } from '@fluentui/react-utilities';
+import type { MessageBarTitleState, MessageBarTitleSlots } from './MessageBarTitle.types';
+
+/**
+ * Render the final JSX of MessageBarTitle
+ */
+export const renderMessageBarTitle_unstable = (state: MessageBarTitleState) => {
+  assertSlots<MessageBarTitleSlots>(state);
+
+  // TODO Add additional slots in the appropriate place
```
Comment (layershifter) [inline packages/react-components/react-message-bar-preview/src/components/MessageBarTitle/renderMessageBarTitle.tsx:13]: Can be removed?

File: packages/react-components/react-message-bar-preview/src/components/MessageBarTitle/renderMessageBarTitle.tsx [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+  return (
+    <>
```
Comment (layershifter) [inline packages/react-components/react-message-bar-preview/src/components/MessageBarTitle/renderMessageBarTitle.tsx:15]: Useless Fragment

**Pull request merged**